src/neo/SmartContract/ApplicationEngine.Contract.cs

```diff
@@ -131,9 +131,16 @@ private void CallContractInternal(UInt160 contractHash, string method, Array arg
             ContractMethodDescriptor md = contract.Manifest.Abi.GetMethod(method);
             if (md is null) throw new InvalidOperationException($"Method {method} Does Not Exist In Contract {contractHash}");
 
-            ContractState currentContract = Snapshot.Contracts.TryGet(CurrentScriptHash);
-            if (currentContract?.CanCall(contract, method) == false)
-                throw new InvalidOperationException($"Cannot Call Method {method} Of Contract {contractHash} From Contract {CurrentScriptHash}");
+            if (md.Safe)
+            {
+                flags &= ~CallFlags.AllowModifyStates;
```

Comment:
If a method is marked as `Safe`, will it not be allowed to modify states?

Reply:
Yes.